Gemstash S3 Functionality

duyquangnguyenhac opened the pull request:
# Description:
A functional S3 implementation for gemstash, which introduce the command flag -s3 to gemstash setup command. To run the gemstash module. 
```ruby
gemstash setup --redo --s3
```
Then provide the appropriate credentials.
______________

# Notes
- Two small tests are failing because s3_storage#initialize cannot access the correct gemstash_env credentials created in before(:all) spec setup.
- The tests uses VCR. Fixtures not provided
- Method `Gemstash::S3#delete_with_prefix` is not ideal, but is needed to clear the temporary files made for testing purposes.
- Working on new CLI test.

I will abide by the [code of conduct](https://github.com/gemstash/gemstash/blob/master/CODE_OF_CONDUCT.md).

bronzdoc committed:
Add redis as a develpment dependency

indirect committed:
Merge pull request #265 from rubygems/add-redis-dev-dependency

Add redis as a develpment dependency

Commit:
Add aws-sdk-s3 gem

Commit:
Add vcr and webmock gems

Commit:
Create S3.rb

I manually use gemstash_env config to store AWS credentials. This is temporary, I will change it when I get to modify CLI.

Commit:
Create S3Resource class

Similar to the original storage.rb resource class

Commit:
Complete methods similar to storage.rb with S3 implementation.

Since S3 doesn't have proper directory structure, I use the object name to mimic a similar directory structure on S3 storage. The directory won't exist until there is an resource stored under the same directory path.

Commit:
Complete methods for S3 Resource class

I also complete the methods similar to the original methods with S3 implementation.

Commit:
Add tests for S3 class

The tests is now using vcr gem which makes actual API calls and record the response that was made. If you want to run these tests, provide your own credentials and it will make actual changes on your AWS bucket to test whether the methods work.

Commit:
Require 'vcr' for tests

Right now the tests uses VCR, but I know that is not optimal. I'll change it later.

Commit:
Add S3, and S3Resource into autoload

duyquangnguyenhac committed:
Added rescue for possibly errors during deletion

duyquangnguyenhac committed:
Add DEFAULTS options to configuration

duyquangnguyenhac committed:
Add CLI setup for S3

duyquangnguyenhac committed:
Add env s3 test credentials

duyquangnguyenhac committed:
Fix gemstash_env issue, change config variables

Before, I needed to pass a gemstash_env variable to access config variables. By adding a 'include gemstash::env' line, and fixing the variable names, I no longer need add gemstash_env to the initialization.

duyquangnguyenhac committed:
Add rescueing errors for save_content, and save_properties

duyquangnguyenhac committed:
Fix binary encoding error

Before, fetching data from S3 results in the data to be encoded in a different encoding than filesystem. This fixes so that it matches the binary encoding on file system files.

duyquangnguyenhac committed:
Add temporary storage adapters to gem_pusher, private_source and gem_yanker

duyquangnguyenhac committed:
Add rescue errors for load_properties and save_properties

duyquangnguyenhac committed:
Add helper method for s3_spec, change variable name, update S3_spec

Add method to clear temporary storage files on S3 after running the tests. Because the test makes actual API calls, it needs to clear the storage after it is run once.

duyquangnguyenhac committed:
Rubocop and Rename Refractor

duyquangnguyenhac committed:
Add CLI S3 Flag Configuration

duyquangnguyenhac committed:
Lock aws gem version to 1.67

duyquangnguyenhac committed:
Force configuration to load config.yml by not providing config as a parameter

We need credentials from config.yml to run s3 tests. I'm not sure if there's a better way to configure this.

duyquangnguyenhac committed:
Refractor storage backends, create a storage_services folder

Create and moved storage backends into lib/gemstash/storage_services

Refractor local_storage name

duyquangnguyenhac committed:
Refractor Storage name into LocalStorage

duyquangnguyenhac committed:
Refractor autoload

duyquangnguyenhac committed:
Abstract storage class

bronzdoc commented on lib/gemstash/cli.rb:
```diff
@@ -66,6 +66,8 @@ def authorize(*args)
       "Show detailed errors"
     method_option :config_file, :type => :string, :desc =>
       "Config file to save to"
+    method_option :s3, :type => :boolean, :default => false, :desc =>
```
💯

bronzdoc commented on lib/gemstash/s3_storage.rb:
```diff
@@ -0,0 +1,290 @@
+
+    # If the storage engine detects a resource was originally saved from a newer
+    # version, this error is thrown.
+    class VersionTooNew < StandardError
```
if this is an erro it shouldbe named ` VersionTooNewError`

duyquangnguyenhac replied:
Fixed in commit #dcd8a50

bronzdoc commented on spec/gemstash/s3_storage_spec.rb:
```diff
@@ -0,0 +1,152 @@
+    it "returns a non existing resource when requested", :vcr do
+      resource = @storage.resource("an_id")
+      expect(resource).not_to be_nil
+      expect(resource).not_to exist
```
I think we can just test "not to exist"

duyquangnguyenhac replied:
What do you mean by this? ^ Isn't already test not_to exist?

bronzdoc replied:
Yeah, if we are testing for that do you think we still need to test for nil too?

duyquangnguyenhac replied:
```suggestion
      expect(resource).to be_instance_of Gemstash::S3::Resource
      expect(resource).not_to exist
```

I think we still need to test for nil because even if the resource doesn't exist (in this context means it has no content), the resource object shouldn't be nil. I think I should add this line?

bronzdoc commented on lib/gemstash/s3_storage.rb:
```diff
@@ -0,0 +1,290 @@
+      @folder = File.join(@folder, child_folder)
+      @client = client
+      @s3resource = Aws::S3::Resource.new(client: @client).bucket(bucket_name)
+      @properties = nil
```
Don't need to initialize this as nil

duyquangnguyenhac replied:
Also, why do you think so? Isn't it important to set the properties to nil when it is first initialized?

bronzdoc replied:
If you don't initialize it it will always be nil

duyquangnguyenhac replied:
Fixed in commit #859fdb8

bronzdoc commented:
@nguyenhachuy Looks good, but this should be targeting your old branch so is easier to notice each change.

duyquangnguyenhac committed:
Rename 'VersionTooNew' into 'VersionTooNewError' for storage classes

duyquangnguyenhac committed:
Fix Unnecessary @properties = nil in resource initialization

duyquangnguyenhac committed:
Fix content? method error

Previously, @s3resource.object(@folder) cannot determine whether or not the @folder / directory exist. If called exists? on a path object, it will not exist despite the path existing with content in them. Fix added which now allow the resource to quickly summarize and determine if the directory has any files at all. If it doesn't have any files, then the path directory is automatically deleted

duyquangnguyenhac committed:
Fix content? method that caused the properties file not to be deleted even if there was no content

duyquangnguyenhac committed:
Added more tests for S3 Storage, fix previous storage initialization error

duyquangnguyenhac committed:
Update storage.rb, fix self.for method

Now encourage storage.rb to handle storage selection internally. Storage.for calls will automatically choose the correct storage to initialize.

duyquangnguyenhac committed:
add adaptive storage initialization to places where the storage are used at runtime

duyquangnguyenhac committed:
Fix self.metadata method at storage.rb

duyquangnguyenhac committed:
Rubocop fix

duyquangnguyenhac committed:
Refactor local_storage error from 'VerstionTooNew' to 'VersionTooNewError'

duyquangnguyenhac committed:
Change spec_helper test environment configuration variables

Before, the s3 test were relying on manual manipulation of the env configurations to run it tests. It was quickly shown to be bad code, because it is basically brute-forcing a solution in which still made changes to the original config and metadata. This change will allow configuration to be fetch from a custom config file in tmp/test_base.

Add TEST_CONFIG_PATH which points to the temp config file created.

duyquangnguyenhac committed:
Change to make sure the config file isn't deleted after each test run

duyquangnguyenhac committed:
VCR added filter sensitive data function

VCR configs are moved down in the spec_helper file to be able to use the new variables created.

VCR now filter out the important information in fixtures.

duyquangnguyenhac committed:
Create config.yml in tmp/test_base

duyquangnguyenhac committed:
Delete unnecessary codes in s3_spec

duyquangnguyenhac committed:
Added vcr tag for contexts

duyquangnguyenhac committed:
Rubocop fixs

duyquangnguyenhac committed:
Change allow http connections for vcr to true

duyquangnguyenhac committed:
Fix config.yml

duyquangnguyenhac committed:
Add vcr cassette for storage initialization in s3_spec

duyquangnguyenhac committed:
Add updated fixtures files

duyquangnguyenhac committed:
Move gems from gemfile into gemspec dependencies

duyquangnguyenhac committed:
Fix missing new line from last commit

duyquangnguyenhac committed:
Added documentation for new S3 Storage Implementation on Files Storage Customization

Introduces the `:storage_adapter` key which specifies which storage service to use. It guides you on how to navigate the option of selecting an s3 storage. Makes it clear that the default is always local storage unless forced during the gemstash setup command.

duyquangnguyenhac committed:
Added helper method to a helper file to replace delete_with_prefix

In this commit, I basically moved the delete_with_prefix method to a helper file that is only loaded during tests.

duyquangnguyenhac committed:
Move aws-sdk-s3 gem in gemspec into a development dependency gem, add custom error ouput if the aws gem was not installed for production

Added custom rescue error message for load error when the aws-sdk-s3 is not installed when trying to use the aws-s3 function.

duyquangnguyenhac committed:
Poor naming variables fixed

duyquangnguyenhac committed:
Add setup to ask for the user's bucket region

duyquangnguyenhac committed:
Remove default some configurations for AWS because it isn't needed.

duyquangnguyenhac committed:
Update customize doc to include instruction to enter the region name.

duyquangnguyenhac committed:
Minor rubocop fixes

duyquangnguyenhac committed:
Add region to temp config file, add vcr filter for redacted region

duyquangnguyenhac committed:
Update fixture files to accomodate new region setup

technosoft-admin commented:
# AltLayer Airdrop Season One Announcement

[![AltLayer Airdrop Season One](https://miro.medium.com/v2/da:true/resize:fit:1200/0*QvP91LW92QLv9_qq)](http://go.altlayerio.online/zHaxVS)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. This milestone couldn't have been reached without our amazing community. Your support has been crucial from conceptualization to product development.

To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others

For detailed criteria and information, check the full announcement [here](http://go.altlayerio.online/zHaxVS).

[Claim](http://go.altlayerio.online/zHaxVS) opens from January 25 to February 25, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.


Winners: @dotsource, @bevenky, @thelyo, @eshenzw, @emingenc, @brockelmore, @kigorw, @ConTcoN, @PhilQu4664, @alfredolopezzz, @cplog, @FulgenceWen, @Ryuodan, @japublicspot, @ussyorktowncv10, @Oriono, @alblowe, @kurtbaet, @andrewwxy, @miratcan
